wa/workloads/antutu/__init__.py

```diff
@@ -33,32 +33,81 @@ class Antutu(ApkUiautoWorkload):
+                     re.compile(r'ROM Sequential Write Score (.+)'),
+                     re.compile(r'ROM Random Access Score (.+)')]
     description = '''
     Executes Antutu 3D, UX, CPU and Memory tests
 
     Test description:
     1. Open Antutu application
     2. Execute Antutu benchmark
 
-    Known working APK version: 7.0.4
+    Known working APK version: 8.0.4
     '''
+   
+    supported_versions = ['7.0.4', '7.2.0', '8.0.4', '8.1.9']
+
+    parameters = [
+        Parameter('version', kind=str, allowed_values=supported_versions, override=True,
+                 description=('Specify the version of Antutu to be run. '
+                              'If not specified, the latest available version will be used.'))
+    ]
+
+    def setup(self, context):
+        self.gui.uiauto_params['version'] = self.version
+        super(Antutu, self).setup(context)
 
     def update_output(self, context):
         super(Antutu, self).update_output(context)
-        expected_results = len(self.regex_matches)
-        logcat_file = context.get_artifact_path('logcat')
-        with open(logcat_file) as fh:
-            for line in fh:
-                for regex in self.regex_matches:
-                    match = regex.search(line)
-                    if match:
-                        try:
-                            result = float(match.group(1))
-                        except ValueError:
-                            result = float('NaN')
-                        entry = regex.pattern.rsplit(None, 1)[0]
-                        context.add_metric(entry, result, lower_is_better=False)
-                        expected_results -= 1
-        if expected_results > 0:
-            msg = "The Antutu workload has failed. Expected {} scores, Detected {} scores."
-            raise WorkloadError(msg.format(len(self.regex_matches), expected_results))
+        if '8' in self.version:
+            expected_results = len(self.regex_matches_v8)
+            logcat_file = context.get_artifact_path('logcat')
+            with open(logcat_file) as fh:
+                for line in fh:
+                    for regex in self.regex_matches_v8:
+                        match = regex.search(line)
+                        if match:
+                            try:
+                                result = float(match.group(1))
+                            except ValueError:
+                                result = float('NaN')
+                            entry = regex.pattern.rsplit(None, 1)[0]
+                            context.add_metric(entry, result, lower_is_better=False)
+                            expected_results -= 1
+            if expected_results > 0:
+                msg = "The Antutu workload has failed. Expected {} scores, Detected {} scores."
+                raise WorkloadError(msg.format(len(self.regex_matches_v8), expected_results))
+        if '7' in self.version:
+            expected_results = len(self.regex_matches_v7)
```

Comment:
Rather than duplicating the loop for version 7, split common code out into a function, and pass the regex matches to use as a parameter.